ibis/expr/rewrites.py

```diff
@@ -297,3 +297,80 @@ def simplify(node):
     node = node.replace(subsequent_projects | subsequent_filters)
     node = node.replace(complete_reprojection)
     return node
+
+
+@replace(p.Comparison)
+def rebase_predicates_on_table_underlying_jointable(_):
+    left, right = _.left, _.right
+    left_table = left.rel.parent
+    right_table = right.rel.parent
+    return _.copy(
+        left=left.copy(rel=left_table, name=left.name),
+        right=right.copy(rel=right_table, name=right.name),
+    )
+
+
+@replace(p.JoinChain)
+def rewrite_join_chain_for_semi_anti_join(_, y=None):
+    # TODO (mehmet): Added optional arg `y` to work around the
+    # error raised for `test_many_subqueries()`.
+    # Is there a cleaner way to address this?
+
+    links = _.find_topmost(ops.JoinLink)
+    semi_anti_links, new_links = [], []
+    for link in links:
+        if link.how in {"semi", "anti"}:
+            semi_anti_links.append(link)
+        else:
+            new_links.append(link)
+
+    # Nothing to do with the original join chain
+    if not semi_anti_links:
+        return _
+
+    if new_links:
+        node = _.copy(rest=tuple(new_links))
+    else:
+        node = _.first.parent
+
+    for link in semi_anti_links:
+        # Note: `filter` is necessary to avoid rebasing predicates
+        # in the `JoinChain`'s deeper down the tree. E.g., when
+        # a table is semi/anti-joined with an inner-join as
+        # table.join(inner_join, predicates=..., how="semi/anti")
+        link = link.replace(
+            rebase_predicates_on_table_underlying_jointable,
+            filter=AnyOf(p.JoinLink | p.Comparison | p.Field | p.JoinTable),
+        )
+
+        # Rebase predicates on `node` if the underlying table for
+        # a predicate lies within the tree starting at `node`.
+        # TODO (mehmet): Is there a cleaner way to do this?
+        # Perhaps using `.replace()`?
+        predicates = []
+        for pred in link.predicates:
+            relations_under_node = node.find(ops.Relation)
+            if pred.left.rel in relations_under_node:
+                pred = pred.copy(left=pred.left.copy(rel=node))
+            if pred.right.rel in relations_under_node:
+                pred = pred.copy(right=pred.right.copy(rel=node))
```

Comment:
Please confirm, the idea here is to replace the fields referencing the join tables under the predicates. This could be done in a way similar to the following:

```py
relation_mapping = {link.table: node for link in new_links}
new_predicate = pred.replace(relation_mapping)
```

Reply:
> Please confirm, the idea here is to replace the fields referencing the join tables under the predicates. 

Yes, this step is meant to "rebase" the fields based on the join tables (`ReferenceTable`'s as newly named) under the predicates.